Adaptive sampling for SSS [skip ci]

EvgeniiG opened the pull request:
### Purpose of this PR
- Added support for Quality Levels to Subsurface Scattering.
- The number of samples is now automatically adjusted at runtime, so there is a chance to use fewer samples than what is budgeted for.
- Also fixed a weird bug with very bright lights (the new test images look more correct).
- Removed a ton of code and shader variables.

---
### Testing status

**Manual Tests**: Opened test project + Run graphic tests locally.
**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline/tree/HDRP%252Fadaptive-sss/.yamato%252Fupm-ci-abv.yml%2523all_project_ci_trunk

---
### Comments to reviewers
I also touched the diffusion profile preview code. The visuals are now better (I think?).

EvgeniiG commented:
Updated ref images.

EvgeniiG commented:
All tests pass except for 1351 and 1710.

sebastienlagarde commented:
@EvgeniiG any thought about performance? does it vary compare to previous version?
@Unity-Technologies/gfx-qa-hdrp Please check the migration with "high quality SSS" as well. thanks

EvgeniiG commented:
I fixed a broken test. Tests should now pass.
I added a migration step for High Quality SSS.
Perf-wise, we'll have to see. I suspect it will be slower as long as the algorithm needs all the samples (num_samples_required = sample_budget) due to the extra ALU. It should be faster for objects at distance (when adaptive sampling kicks in). I will run some tests.

EvgeniiG commented:
![image](https://user-images.githubusercontent.com/23363056/75833800-0089b800-5d6f-11ea-82d3-98e7632364bb.png)

EvgeniiG commented:
I did a little bit of testing on the PS4. I compared the current version with the version where I manually pass the constants to the shader and tell the compiler to unroll the loop (so it can importance-sample at the compile time).

The shader is heavily memory-bound, so the extra ALU for importance sampling only makes around 3-5% difference.

I should note that it is still very easy to destroy the perf by using a large scattering distance (or simply jamming the camera into the character's face), which will result in trashing the texture cache and cost of 10 ms+ even on low settings.

EvgeniiG commented:
It's 6 wavefronts per SIMD, used to be 7.

![image](https://user-images.githubusercontent.com/23363056/75932484-58d2bf80-5e2c-11ea-8667-e456d5b5e058.png)

EvgeniiG commented:
I will see if I can win another wavefront by massaging it in the SCUI.

EvgeniiG commented:
7x wavefronts now, not really faster, though.

![image](https://user-images.githubusercontent.com/23363056/75936815-1adb9880-5e38-11ea-9e05-179650a611cd.png)

EvgeniiG commented:
Razor's data for the SSS pass looks pretty healthy, so not sure what else we can do here except reduce the number of samples.

EvgeniiG commented:
This is 2.5 ms.

![image](https://user-images.githubusercontent.com/23363056/75937895-0fd63780-5e3b-11ea-9054-5b4ab661bddb.png)

FrancescoC-unity commented on com.unity.render-pipelines.high-definition/Runtime/Lighting/SurfaceShading.hlsl:
```diff
@@ -132,24 +132,18 @@ float3 EvaluateTransmittance_Punctual(LightLoopContext lightLoopContext,
-    float3 p = thicknessDelta * S;
-    float3 expOneThird = exp2(p);
-#endif
+    float3 exp_13 = exp2(((LOG2_E * (-1.0/3.0)) * dt) * S); // Exp[-S * r / 3]
```
Should this comment be // Exp[-S * dt / 3] ?

EvgeniiG replied:
Indeed, although there's no 'r' in this code.

FrancescoC-unity commented on com.unity.render-pipelines.high-definition/Runtime/Material/DiffusionProfile/DiffusionProfile.hlsl:
```diff
@@ -27,14 +12,47 @@ float3 ComputeTransmittanceDisney(float3 S, float3 volumeAlbedo, float thickness
+float3 EvalBurleyDiffusionProfile(float r, float3 S)
+{
+    float3 exp_13 = exp2(((LOG2_E * (-1.0/3.0)) * r) * S); // Exp[-S * r / 3]
+    float3 expSum = exp_13 * (1 + exp_13 + exp_13);        // Exp[-S * r / 3] + Exp[-S * r]
```
To get what the comment says I think it should be exp_13 * (1 + exp_13 ***** exp_13); and not plus, is the comment or code correct?

EvgeniiG replied:
Oh crap, that's a typo. Good catch.

FrancescoC-unity commented on com.unity.render-pipelines.high-definition/Runtime/Material/SubsurfaceScattering/SubsurfaceScattering.compute:
```diff
@@ -7,21 +7,19 @@
-#pragma kernel SubsurfaceScatteringMQ_MSAA SubsurfaceScattering=SubsurfaceScatteringMQ_MSAA SSS_ENABLE_NEAR_FIELD=0 ENABLE_MSAA
-#pragma kernel SubsurfaceScatteringHQ_MSAA SubsurfaceScattering=SubsurfaceScatteringHQ_MSAA SSS_ENABLE_NEAR_FIELD=1 ENABLE_MSAA
+#pragma kernel SubsurfaceScattering SubsurfaceScattering=SubsurfaceScattering
+#pragma kernel SubsurfaceScattering_MSAA SubsurfaceScattering=SubsurfaceScattering_MSAA ENABLE_MSAA
```
Small heads up here, since very recently (I think PR got merged today), we support multi_compile in compute. 

I had ported the old SSS to it so it has the multi_compile for ENABLE_MSAA already defined, I guess though that some conflict will come when merging due to the SSS_ENABLE_NEAR_FIELD removal

FrancescoC-unity commented on com.unity.render-pipelines.high-definition/Runtime/Material/SubsurfaceScattering/SubsurfaceScattering.compute:
```diff
@@ -113,17 +110,23 @@ float4 LoadSample(int2 pixelCoord, int2 cacheOffset)
     int2 cacheCoord = pixelCoord - cacheOffset;
     bool isInCache  = max((uint)cacheCoord.x, (uint)cacheCoord.y) < TEXTURE_CACHE_SIZE_1D;
 
-    UNITY_BRANCH if (isInCache)
```
Are all the UNITY_BRANCH  removed because we are good with the compiler choosing for us or because ineffective (compiler not taking the hint)?

EvgeniiG replied:
The compiler will figure it out. Let's not pollute the code.

FrancescoC-unity commented on com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/Settings/FrameSettings.Migration.cs:
```diff
@@ -416,5 +416,16 @@ internal static void MigrateToSeparateColorGradingAndTonemapping(ref FrameSettin
+        internal static void MigrateSubsurfaceParams(ref FrameSettings fs, bool previouslyHighQuality)
+        {
+            // SSS moved from 25 to 46.
+            fs.SetEnabled((FrameSettingsField)25, false);
```
I wonder what happens if we reuse 25 before this package is released or if a user upgrade to when we use 25 again - downgrade - reupgrade.

To be on the safe side we could probably reserve out the slot 25 (adding a comment about it), though I am not 100% sure if this would be ever a problem.

sebastienlagarde commented:
Note: for the conflict on screenshots, the screenshots need to be move as path have change, there is now an extra "None" folder since the PR from Sophia

EvgeniiG commented:
> Favro card: http://go/ihglm
> Two issues so far:
> 
> * Sample budget is being taken from the default HDRP asset only. Switching different quality settings has no effect. (detailed video in favro)
> * In Frame Settings SSS Quality Mode has a toggle check box by default. Should only be visible when you override global SSS toggle.
>   ![image](https://user-images.githubusercontent.com/50581102/77159680-79e31380-6aae-11ea-85bd-c11d08d11a38.png)

Fixed both issues.

EvgeniiG commented:
Tests pass locally.
